<a href="https://colab.research.google.com/github/Luisadle/Business-Intelligence/blob/main/Reto_1_Luisa_De_La_Espriella.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se cargan las librerias necesarias

In [ ]:
import numpy as np
import pandas as pd

1. Conexión con la base de datos

In [ ]:
nxl='/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl, sheet_name=0)
XDB=XDB.dropna() #Evitar celdas vacias
print(XDB)


          ID  Age  Urban  Mother_Edu  Father_Edu  Siblings  Black  Hispanic  \
1          2   20    1.0         5.0         8.0       8.0    0.0       0.0   
5          6   19    1.0        12.0        12.0       1.0    0.0       0.0   
6          7   15    1.0        12.0        12.0       1.0    0.0       0.0   
7          8   21    1.0         9.0         6.0       7.0    0.0       0.0   
8          9   16    1.0        12.0        10.0       4.0    0.0       0.0   
...      ...  ...    ...         ...         ...       ...    ...       ...   
12557  12558   20    1.0        12.0        12.0       6.0    1.0       0.0   
12565  12566   20    1.0        12.0        11.0       2.0    1.0       0.0   
12647  12648   21    1.0        16.0        12.0       0.0    1.0       0.0   
12651  12652   21    1.0        16.0        16.0       2.0    0.0       1.0   
12678  12679   21    0.0         8.0         8.0      10.0    1.0       0.0   

       White  Christian  ...  Height  Weight  Outgo

2. Se seleccionan las variables de interés

In [ ]:
XDB=XDB[['Age','Urban','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]

#Separamos las variables de interés en variables de entradas y variables de salida
#Variables de Entrada
X=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight', 'Income']]
print(X)

#Variable de Salida - Variable Pronóstico
Y=XDB[['Urban']]
print(Y)

       Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  
1          0.0  
5      40000.0  


3. Se procede a la implementación del modelo

In [ ]:
from sklearn.naive_bayes import GaussianNB
mnb=GaussianNB() #Aloja el modelo
mnb.fit(X,Y) #Entrena el modelo

#Obtenemos la información del modelo
u=mnb.theta_ #Obtenemos la media de las variables
print("La media de la varible es:\n", u)

var=mnb.var_ #Obtenemos la varianza de las variables
print("La varianza de la variable es:\n", var)

sigma=np.sqrt(var) #Obtenemos la desviación estándar de las variables
print("La desviación estándar de la variable es:\n", sigma)

#Se determinan los limites superiores e inferiores de las variables
LS=u+sigma
LI=u-sigma
print("El límite superior de la variable es:\n", LS)
print("El límite inferior de la variable es:\n", LI)

#Evaluamos un solicitante
XDS=[[34,12,15,2,24,69,160,48000]] #Como un valor de entrada
ydp=mnb.predict(XDS) #Se obtiene la predicción
print("La predicción del solicitante es:\n", ydp)

if ydp==0:
  print("El solicitante vive en áreas Rurales")
else:
  print("El solicitante vive en áreas Urbanas")

La media de la varible es:
 [[1.83810742e+01 1.07612958e+01 1.03324808e+01 4.74168798e+00
  2.20272805e+01 6.72148338e+01 1.48857630e+02 3.03672506e+04]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 4.68780488e+00
  2.26004435e+01 6.69031042e+01 1.44760976e+02 3.34278929e+04]]
La varianza de la variable es:
 [[6.15727948e+00 1.00639316e+01 1.49720047e+01 6.63304385e+00
  1.75796726e+01 1.66288925e+01 9.06652219e+02 1.00501103e+09]
 [6.12994838e+00 1.14704244e+01 1.68556528e+01 5.94306906e+00
  1.75141489e+01 1.77519886e+01 9.05487615e+02 1.30488762e+09]]
La desviación estándar de la variable es:
 [[2.48138660e+00 3.17237004e+00 3.86936748e+00 2.57546964e+00
  4.19281202e+00 4.07785391e+00 3.01106662e+01 3.17019089e+04]
 [2.47587326e+00 3.38680150e+00 4.10556364e+00 2.43784107e+00
  4.18499090e+00 4.21331089e+00 3.00913213e+01 3.61232282e+04]]
El límite superior de la variable es:
 [[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.31715762e+00
  2.62200925e+01 7.12926877e+01 1.78968296e

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


3. Evaluamos el desempeño del modelo

In [ ]:
from sklearn.metrics import confusion_matrix

#haga un barrido de todos los datos
ydp2=mnb.predict(X) #Haga el pronóstico para los datos base
cm=confusion_matrix(Y,ydp2) #Cree la matriz de confusión
print("La matriz de confusión es:\n", cm)
VN=cm[0,0] #Verdadero Negativo
FP=cm[0,1] #Falso Positivo
FN=cm[1,0] #Falso Negativo
VP=cm[1,1] #Verdadero Positivo

#Metricas de Desempeño
Exactitud=(VP+VN)/(VP+VN+FP+FN) #Comportamiento general PreAbrobando. Por encima del 75% es un buen modelo.
print("La exactitud del modelo es:\n", Exactitud)

TE= (FN+VP)/(VP+VN+FP+FN)
print("La tasa de error  es:\n", TE)

Sensibilidad=VP/(VP+FN) #La aprobación de los positivos, unicamente.
                        #Sobre el total de los positivos
print("La sensibilidad del modelo es:\n", Sensibilidad)

Especificidad=VN/(VN+FP) #Comportamiento correcto negativo
                        #Sobre el total de los negativos
print("La especificidad del modelo es:\n", Especificidad)

Precision=VP/(VP+FP) #Comportamiento correcto positivo
                    #Sobre el total del pronóstico
print("La precisión del modelo es:\n", Precision)

PrNeg=VN/(VN+FN) #Comportamiento correcto negativo
                    #Sobre el total del pronóstico
print("La probabilidad de Negación del modelo es:\n", PrNeg)

La matriz de confusión es:
 [[   2 1171]
 [   4 4506]]
La exactitud del modelo es:
 0.7932430054548654
La tasa de error  es:
 0.7935949322540912
La sensibilidad del modelo es:
 0.9991130820399113
La especificidad del modelo es:
 0.0017050298380221654
La precisión del modelo es:
 0.793729082261758
La probabilidad de Negación del modelo es:
 0.3333333333333333


In [ ]:
totaldatos=len(XDB) #Cantidad de datos en la base de datos
print("El total de datos es:\n", totaldatos)

cantidadV = Y['Urban'].value_counts() #Frecuencia de la variable
print("La cantidad de datos en la variable es:\n", cantidadV)

pv = cantidadV/len(Y) *100 #Probabilidad de la variable
print("La probabilidad de la variable es:\n", pv)

El total de datos es:
 5683
La cantidad de datos en la variable es:
 Urban
1.0    4510
0.0    1173
Name: count, dtype: int64
La probabilidad de la variable es:
 Urban
1.0    79.359493
0.0    20.640507
Name: count, dtype: float64


**Análisis de Resultados:** De acuerdo con la bases de datos suministrada, se tomaron como referencia para la identificación de la vivienda de las personas, las siguientes variables: Edad, Educación de la madre y del padre, el tamaño de la familia, la escala del autoestima, peso, estatura y ingresos. La base de datos posee un total de 5683 datos, los cuales se dividieron en un 79% aproximadamente de personas que viven en áreas urbanas, lo que equivale a 4510 personas, asimismo hubo un 20% aproximadamente de personas que viven en áreas rurales, el cual equivale a 1173 personas.

Por otro lado, respecto a las variables utilizadas para lograr este análisis, podemos establecer que las edades se encontraban en un intervalo de 15 a 22 años, adicionalmente la estatura de las personas fluctuan entre 48 y 83, asimismo el peso de las personas se encuentran en un intervalo de 53 y 375.

Es importante destacar que el modelo presento una exactitud cercana al 80% (79%), lo cual demuestra un comportamiento aceptable para la toma de decisiones. Igualmente el modelo se destaco de manera casi perfecta al analizar la sensibilidad, ya que este dio como resultado 99,91%, demostrando un excepcional conducta para identificar el área de vivienda de las personas. Por último, el modelo presento una especificidad de aproximadamente 0,17%.

Teniendo en cuenta los datos proporcionados de las variables de entradas, mencionadas durante el desarrollo del la actividad. 34,12,15,2,24,69,160,48000, siguiendo el orden de las variables ('Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight', 'Income') Logramos concluir que esta persona vive en área urbana, es decir en la ciudad.
